src/mdns/mdns_mdnssd.cpp

```diff
@@ -408,131 +409,224 @@ void PublisherMDnsSd::HandleServiceRefDeallocating(const DNSServiceRef &aService
+void PublisherMDnsSd::DnssdServiceRegistration::HandleRegisterResult(DNSServiceFlags aFlags, DNSServiceErrorType aError)
+{
+    if ((aError == kDNSServiceErr_NoError) && (aFlags & kDNSServiceFlagsAdd))
     {
-        const DNSRecordRef &recordRef = recordRefAndAddress.first;
-        const Ip6Address   &address   = recordRefAndAddress.second;
-        if (IsCompleted())
-        {
-            // The Bonjour mDNSResponder somehow doesn't send goodbye message for the AAAA record when it is
-            // removed by `DNSServiceRemoveRecord`. Per RFC 6762, a goodbye message of a record sets its TTL
-            // to zero but the receiver should record the TTL of 1 and flushes the cache 1 second later. Here
-            // we remove the AAAA record after updating its TTL to 1 second. This has the same effect as
-            // sending a goodbye message.
-            // TODO: resolve the goodbye issue with Bonjour mDNSResponder.
-            dnsError = DNSServiceUpdateRecord(mServiceRef, recordRef, kDNSServiceFlagsUnique, sizeof(address.m8),
-                                              address.m8, /* ttl */ 1);
-            otbrLogResult(DNSErrorToOtbrError(dnsError), "Send goodbye message for host %s address %s: %s",
-                          MakeFullHostName(mName).c_str(), address.ToString().c_str(), DNSErrorToString(dnsError));
-        }
-        dnsError = DNSServiceRemoveRecord(mServiceRef, recordRef, /* flags */ 0);
-        otbrLogResult(DNSErrorToOtbrError(dnsError), "Remove record for host %s address %s: %s",
-                      MakeFullHostName(mName).c_str(), address.ToString().c_str(), DNSErrorToString(dnsError));
-        // TODO: ?
-        // DNSRecordRefDeallocate(recordRef);
+        otbrLogInfo("Successfully registered service %s.%s", mName.c_str(), mType.c_str());
+        Complete(OTBR_ERROR_NONE);
+    }
+    else
+    {
+        otbrLogErr("Failed to register service %s.%s: %s", mName.c_str(), mType.c_str(), DNSErrorToString(aError));
+        GetPublisher().RemoveServiceRegistration(mName, mType, DNSErrorToOtbrError(aError));
     }
-
-exit:
-    return;
 }
 
-Publisher::ServiceRegistration *PublisherMDnsSd::FindServiceRegistration(const DNSServiceRef &aServiceRef)
+otbrError PublisherMDnsSd::DnssdHostRegistration::Register(void)
 {
-    ServiceRegistration *result = nullptr;
+    DNSServiceErrorType dnsError = kDNSServiceErr_NoError;
 
-    for (auto &kv : mServiceRegistrations)
+    otbrLogInfo("Registering new host %s", mName.c_str());
+
+    for (const Ip6Address &address : mAddresses)
     {
-        // We are sure that the service registrations must be instances of `DnssdServiceRegistration`.
-        auto &serviceReg = static_cast<DnssdServiceRegistration &>(*kv.second);
+        DNSRecordRef recordRef = nullptr;
 
-        if (serviceReg.GetServiceRef() == aServiceRef)
-        {
-            result = kv.second.get();
-            break;
-        }
+        dnsError = GetPublisher().CreateSharedHostsRef();
```

Comment:
I wonder why `mHostsRef` needs to be created for every address? IIUC it's never cleared.

Reply:
Never mind. I just realized `CreateSharedHostsRef` returns no error when the ref is already created.

Reply:
en, so should it be renamed to `GetSharedHostsRef()`? `GetXxxx()` seems more like a singleton

Reply:
I personally try to use `Get{Item}()` for simple getters, and find/use a different verb for anything that does maybe more work than a simple getter.